src/process_utility.c

```diff
@@ -1210,96 +1281,308 @@ process_index_chunk(Hypertable *ht, Oid chunk_relid, void *arg)
-								cmd->type)));
-				break;
-		}
+	ts_with_clause_filter(stmt->options,
+						  HYPERTABLE_NAMESPACE,
+						  &hypertable_options,
+						  &postgres_options);
+
+	stmt->options = postgres_options;
+
+	parsed_with_clauses = ts_with_clauses_parse(hypertable_options,
+												index_with_clauses,
+												TS_ARRAY_LEN(index_with_clauses));
+
+	info.extended_options.multitransaction =
+		DatumGetBool(parsed_with_clauses[HypertableIndexFlagMultiTransaction].parsed);
+#ifdef DEBUG
+	info.extended_options.max_chunks =
+		DatumGetInt32(parsed_with_clauses[HypertableIndexFlagMaxChunks].parsed);
+	info.extended_options.barrier_table =
+		DatumGetObjectId(parsed_with_clauses[HypertableIndexFlagBarrierTable].parsed);
+#endif
+
+	/* Make sure this index is allowed */
+	if (stmt->concurrent)
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("hypertables do not support concurrent "
+						"index creation")));
 
+	if (info.extended_options.multitransaction &&
+		(stmt->unique || stmt->primary || stmt->isconstraint))
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("cannot use hypertable.multi_transaction with UNIQUE or PRIMARY KEY")));
+
+	ts_indexing_verify_index(ht->space, stmt);
+
+	if (info.extended_options.multitransaction)
+		PreventInTransactionBlock(true, "CREATE INDEX ... WITH (hypertable.multi_transaction)");
+
+	/* CREATE INDEX on the root table of the hypertable */
+	root_table_index = ts_indexing_root_table_create_index(stmt, queryString, info.extended_options.multitransaction);
+	info.obj.objectId = root_table_index.objectId;
+
+	/* CREATE INDEX on the chunks */
+
+	/* create chunk indexes using the same tranaction for all the chunks */
+	if (!info.extended_options.multitransaction)
+	{
+		CatalogSecurityContext sec_ctx;
 		/*
 		 * Change user since chunk's are typically located in an internal
-		 * schema and chunk indexes require metadata changes
+		 * schema and chunk indexes require metadata changes. In the
+		 * multi-tranaction case, we do this once per chunk.
 		 */
 		ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
-
 		/* Recurse to each chunk and create a corresponding index */
 		foreach_chunk(ht, process_index_chunk, &info);
 
 		ts_catalog_restore_user(&sec_ctx);
-		handled = true;
+		ts_cache_release(hcache);
+		return true;
 	}
 
+	/* create chunk indexes using a seperate transaction for each chunk */
+
+	/* we're about to release the hcache so store the main_table_relid for later */
+	main_table_id = ht->main_table_relid;
+	main_table_relation = relation_open(ht->main_table_relid, AccessShareLock);
+	main_table_desc = RelationGetDescr(main_table_relation);
+
+	main_table_index_relation = relation_open(info.obj.objectId, AccessShareLock);
+	main_table_index_lock_relid = main_table_index_relation->rd_lockInfo.lockRelId;
+
+	info.extended_options.indexinfo = BuildIndexInfo(main_table_index_relation);
```

Comment:
is all of this stuff allocated in the correct memory context? i.e. it cannot be in the transaction context? We should probably create a special context for this and delete after?

Reply:
yes, all of the stuff from here that is passed to the per-index transactions is allocated on the stack